### PR TITLE
Update Tor module to add option to encrypt Tor key

### PR DESCRIPTION
**Pull request by orbitalturtle**
This PR lays the groundwork for #6500, which adds the ability to encrypt a Tor private key on disk. We originally had this all in one PR, but needed to split out this refactoring up because changes to the Tor module need to be added in another PR. 

Summary of Changes:

- Lays groundwork in Tor package for encrypting private key

### CODE REVIEW

File: tor/onionfile/onion_file.go
Comment (tvolk131) [inline tor/onionfile/onion_file.go:49]: nit: `OnionFile` -> `File`

File: tor/onionfile/onion_file_test.go
Comment (tvolk131) [inline tor/onionfile/onion_file_test.go:12]: Any reason this isn't just `hide_me_plz` like it was before?

File: tor/onionfile/onion_file_test.go
Comment (tvolk131) [inline tor/onionfile/onion_file_test.go:29]: nit: Change `OnionFile` to either `onionFile` to match the package name, or `File` to match the new struct name.

**Review by tvolk131 — CHANGES_REQUESTED** [review]
Looks good! Just a few nitpicky things that need to be cleaned up.

File: tor/onionfile/onion_file_test.go
Comment (orbitalturtle) [reply tor/onionfile/onion_file_test.go:12]: Ah yup so that's added as a way to tell if the key is encrypted or not, see this line https://github.com/lightningnetwork/lnd/pull/6526/files#diff-a219f7652057175a72ef2276ada3396923cfd0c3e82b4149bd1705cb2a70a9d5R111

**Comment by orbitalturtle** [comment]
Great, thanks for the review @tvolk131 ! Made those changes.

File: tor/onionfile/onion_file.go
Comment (tvolk131) [inline tor/onionfile/onion_file.go:105]: I think this chunk can be made a bit easier to read:

```
if _, err := os.Stat(f.privateKeyPath); errors.Is(err, os.ErrNotExist) {
	return nil, ErrNoPrivateKey
}

privateKeyContent, err := ioutil.ReadFile(f.privateKeyPath)
if err != nil {
	return nil, err
}

// If the privateKey doesn't start with either v2 or v3 key params
// it's likely encrypted.
isEncrypted := !bytes.HasPrefix(privateKeyContent, []byte(V2KeyParam)) &&
	!bytes.HasPrefix(privateKeyContent, []byte(V3KeyParam))

if isEncrypted {
	// If the privateKeyContent is encrypted but --tor.encryptkey
	// wasn't set we return an error
	if !f.encryptKey {
		return nil, ErrEncryptedTorPrivateKey
	}
	// Attempt to decrypt the key
	reader := bytes.NewReader(privateKeyContent)
	privateKeyContent, err = f.encrypter.DecryptPayloadFromReader(reader, f.keyRing)
	if err != nil {
		return nil, err
	}
}
return privateKeyContent, nil
```

File: tor/onionfile/onion_file_test.go
Comment (tvolk131) [reply tor/onionfile/onion_file_test.go:12]: Gotcha, makes sense!

**Review by tvolk131 — APPROVED** [review]

**Comment by tvolk131** [comment]
LGTM! I left one comment about a potential cleanup.

**Comment by orbitalturtle** [comment]
Rebased this just FYI :)

File: tor/onionfile/onion_file.go
Comment (Crypt-iQ) [inline tor/onionfile/onion_file.go:105]: nit: can avoid indenting this block by exiting early if the file isn't found?

File: tor/onionfile/onion_file.go
Comment (Crypt-iQ) [inline tor/onionfile/onion_file.go:114]: not sure how the encryption is done, but is it possible that somebody is unlucky and their encrypted key matches the prefix?

File: tor/cmd_onion.go
Comment (Crypt-iQ) [inline tor/cmd_onion.go:27]: It is a little odd that we're checking that `onionfile.File` (which is in a different package) adheres to this package's `OnionStore` interface. Normally they would live in the same package.

File: tor/onionfile/onion_file_test.go
Comment (Crypt-iQ) [inline tor/onionfile/onion_file_test.go:38]: need `defer os.RemoveAll(tempDir)`

**Review by Crypt-iQ — CHANGES_REQUESTED** [review]
Looking at the PR that depends on this, it doesn't seem like there's a need for the `onionfile` pkg. I think this is mostly there

File: tor/onionfile/onion_file.go
Comment (Crypt-iQ) [reply tor/onionfile/onion_file.go:114]: i think this is like a 1/2^56 chance so i think it's ok

File: tor/cmd_onion.go
Comment (Crypt-iQ) [reply tor/cmd_onion.go:27]: I believe the `OnionStore` iface can actually be unexported, but the methods exported. This lets people know that it can't be used outside of the `tor` package. This is what we do in the funding manager (https://github.com/lightningnetwork/lnd/blob/cf9a9864cf253dbbcac5904d360bbbde763e1ebe/funding/manager.go#L476)

File: tor/cmd_onion.go
Comment (guggero) [inline tor/cmd_onion.go:91]: nit: break between parameters or return types, not before parenthesis.

File: tor/onionfile/onion_file.go
Comment (guggero) [inline tor/onionfile/onion_file.go:62]: nit: break between parameters, not before parenthesis.

File: tor/onionfile/onion_file.go
Comment (guggero) [inline tor/onionfile/onion_file.go:78]: set `privateKeyContent := privateKey` and move both `b` and `payload` into the `if`?
Then we don't need an `else` statement.

File: tor/onionfile/onion_file.go
Comment (guggero) [reply tor/onionfile/onion_file.go:114]: The key params are strings of 5 or more characters, very very unlikely to ever get that from binary data at random.

File: tor/onionfile/onion_file.go
Comment (guggero) [inline tor/onionfile/onion_file.go:115]: nit: add empty line after multi-line `if` statement.
Also, multiple lines that are too long (use 8 spaces for tab character) and multiple comments without punctuation.

File: tor/onionfile/onion_file.go
Comment (guggero) [inline tor/onionfile/onion_file.go:113]: Invert the `if` condition and return early to avoid indentation?

File: tor/cmd_onion.go
Comment (guggero) [inline tor/cmd_onion.go:15]: If the onion type is never actually used in these methods (and we're going to remove V2 soon anyway), should we remove them in this refactor?

File: tor/onionfile/onion_file.go
Comment (guggero) [inline tor/onionfile/onion_file.go:81]: This commit does more than just moving code around. I think we should adjust the commit message accordingly.
I think for this amount of code it's okay to keep it in one commit, but normally we'd follow this pattern:
 1. A commit that purely moves the code and updates other components to use the moved code (e.g. updates to package names)
 2. A commit that removes outdated functionality (e.g. the `OnionType`)
 3. A commit that adds new functionality

File: tor/onionfile/onion_file_test.go
Comment (guggero) [reply tor/onionfile/onion_file_test.go:38]: Instead, we should start using `t.TempDir()` everywhere in tests, that automatically removes the directories.

File: tor/onionfile/onion_file_test.go
Comment (guggero) [inline tor/onionfile/onion_file_test.go:49]: Use `require.ErrorIs()` here and `require` everywhere else instead of `t.Fatalf()`.

File: tor/onionfile/onion_file_test.go
Comment (guggero) [inline tor/onionfile/onion_file_test.go:87]: Can use `require.Equal()`.

**Review by guggero — COMMENTED** [review]
I know this PR (or its predecessor) is quite old, so we changed a few things along the line that require an update in the code.
But the main functionality looks pretty good, so there shouldn't be any major changes necessary and we should probably be able to get this in soon.

File: tor/cmd_onion.go
Comment (orbitalturtle) [reply tor/cmd_onion.go:27]: Makes sense!

File: tor/cmd_onion.go
Comment (orbitalturtle) [reply tor/cmd_onion.go:15]: Yeah if they're not used anyway, makes sense to me

Comment (orbitalturtle) [reply tor/cmd_onion.go:15]: Took a stab at removing these

File: tor/onionfile/onion_file.go
Comment (orbitalturtle) [reply tor/onionfile/onion_file.go:100]: Makes sense. I added these details to the commit message, but will keep this in mind in the future

**Comment by orbitalturtle** [comment]
Thanks for the reviews @Crypt-iQ and @guggero! I made those changes as well as rebased the PR.

File: tor/cmd_onion.go
Comment (Crypt-iQ) [inline tor/cmd_onion.go:27]: You should be able to remove the exported `OnionStore` now and replace it with `onionStore`. If I'm wrong then you can just get rid of the unexported `onionStore`

**Review by Crypt-iQ — COMMENTED** [review]
Almost there. I don't see the point of the `onionfile` directory - am I missing something? It's not a package so I think that anything that imports it also imports the `tor` package so there wouldn't be a separation b/t the `tor` and `tor/onionfile`.

**Comment by orbitalturtle** [comment]
@Crypt-iQ Ah so the original intent of moving onionfile comes from this conversation from way back when: https://github.com/lightningnetwork/lnd/pull/4458#discussion_r620729177
But perhaps since the Tor directory has evolved into being a module instead of a package, this no longer makes sense? Will defer to you and @guggero on that one.

**Comment by guggero** [comment]
> @Crypt-iQ Ah so the original intent of moving onionfile comes from this conversation from way back when: [#4458 (comment)](https://github.com/lightningnetwork/lnd/pull/4458#discussion_r620729177) But perhaps since the Tor directory has evolved into being a module instead of a package, this no longer makes sense? Will defer to you and @guggero on that one.

Ah, I was trying to remember why we decided to move the file in the first place. I knew there was a reason but couldn't recall. I think the `EncrypterDecrypter` interface decouples things sufficiently so we don't have a dependency to `lnencrypt`. Therefore we can probably revert the move. Sorry about the back and forth...

**Comment by orbitalturtle** [comment]
@guggero Ahh that's right, I forgot we had completely removed lnencrypt as a dependency. No worries! Updated the code to remove the new onionfile package and I updated the PR description, so this is ready for another look.

File: tor/cmd_onion_test.go
Comment (Crypt-iQ) [inline tor/cmd_onion_test.go:230]: nit: can remove newline

**Review by Crypt-iQ — APPROVED** [review]
lgtm pending nit fix

File: tor/cmd_onion.go
Comment (guggero) [inline tor/cmd_onion.go:147]: nit: add empty line before comment.

File: tor/cmd_onion.go
Comment (guggero) [inline tor/cmd_onion.go:320]: nit: missing punctuation.

File: tor/cmd_onion_test.go
Comment (guggero) [inline tor/cmd_onion_test.go:18]: nit: missing punctuation.

File: tor/cmd_onion.go
Comment (guggero) [inline tor/cmd_onion.go:62]: I think instead of `bytes.Buffer` this should just be `[]byte`.

File: tor/cmd_onion_test.go
Comment (guggero) [inline tor/cmd_onion_test.go:231]: Should we return another string here so we can test that the encrypter was actually called?

File: tor/onionaddr.go
Comment (guggero) [inline tor/onionaddr.go:49]: nit: this comment is no longer accurate, we're now always returning the private key.

**Review by guggero — COMMENTED** [review]
Thank you for the updates!
A few last small comments, after those we're good to merge this :tada:

File: tor/cmd_onion.go
Comment (orbitalturtle) [reply tor/cmd_onion.go:62]: Yeah that makes sense as we need to convert it to bytes in the current function implementation anyway. I'll update this here and in the other PR as well

File: tor/cmd_onion_test.go
Comment (orbitalturtle) [reply tor/cmd_onion_test.go:231]: Another string?

Comment (guggero) [reply tor/cmd_onion_test.go:227]: Something like `this_is_the_encrypted_key` instead of the normal private key. Then in the test make sure the "encrypted" key is returned instead of the plaintext `privateKey`.

Comment (orbitalturtle) [reply tor/cmd_onion_test.go:227]: Ahhh I see. That makes sense. I feel iffy about calling it an encrypted key since it's being returned by the decrypter, so I made one called "anotherKey." Otherwise, does the new change look like what you had in mind?

**Review by guggero — APPROVED** [review]
LGTM :tada:

**Comment by guggero** [comment]
Pushed a new tag `tor/v1.1.0` with the merge commit.

**Pull request merged**
